### PR TITLE
fix(http): Avoid direct URL.resolve calls

### LINKED ISSUE #7317: Avoid URL.resolve()
**What Renovate type, platform and version are you using?**

Any

**Describe the bug**

`URL.resolve('https://host.com/basepath/', '/api/dep.json') === 'https://host.com/api/dep.json'`, which most of the time is not OK.

On the other hand, `URL.resolve('https://host.com/basepath/', 'https://otherhost.com/api/dep.json') === 'https://otherhost.com/api/dep.json'` which is what we want.

We should check every use of url.resolve and see if it can be replaced with the `url-join` package. We perhaps need to make sure that the right-hand side path is not a full URL itself though.

Fixing in the `util/http` layer is most important.

### PR DESCRIPTION
**Pull request by zharinov**
Closes #7317

### CODE REVIEW

File: lib/util/http/github.ts [hunk 1]
```diff
@@ -161,7 +161,7 @@ export class GithubHttp extends Http<GithubHttpOptions, GithubHttpOptions> {
 
     const method = opts.method || 'get';
 
-    if (method.toLowerCase() === 'post' && url === 'graphql') {
+    if (method.toLowerCase() === 'post' && url === '/graphql') {
```
Comment (zharinov) [inline lib/util/http/github.ts:164]: Don't like this part, I'll think how to improve it

File: lib/util/http/github.spec.ts [hunk 2]
```diff
@@ -43,15 +43,15 @@ describe(getName(__filename), () => {
     it('strips v3 for graphql', async () => {
       httpMock
         .scope('https://ghe.mycompany.com')
-        .post('/graphql')
+        .post('/api/graphql')
```
Comment (rarkins) [inline lib/util/http/github.spec.ts:46]: Wrong?

File: lib/util/http/github.spec.ts [hunk 3]
```diff
@@ -43,15 +43,15 @@ describe(getName(__filename), () => {
...
         .reply(200, {});
       setBaseUrl('https://ghe.mycompany.com/api/v3/');
       await githubApi.postJson('/graphql', {
         body: {},
       });
       const [req] = httpMock.getTrace();
       expect(req).toBeDefined();
-      expect(req.url).not.toContain('/v3');
+      expect(req.url).toEqual('https://ghe.mycompany.com/api/graphql');
```
Comment (rarkins) [inline lib/util/http/github.spec.ts:54]: I thought it's meant to be /graphql not /api/graphql

**Comment by rarkins** [comment]
BTW:

image

**Comment by zharinov** [comment]
So, `URL.resolve` output depends on head/tail slashes. What we'll do here is to make sure `util/http` calls do not care about trailing slashes for `baseUrl` as well as leading slashes for `url` (`path`) parameters.

File: lib/platform/bitbucket/__snapshots__/index.spec.ts.snap [hunk 4]
```diff
@@ -1051,7 +1051,7 @@ Array [
       "user-agent": "https://github.com/renovatebot/renovate",
     },
     "method": "GET",
-    "url": "https://api.bitbucket.org/2.0/repositories/?role=contributor&pagelen=100",
+    "url": "https://api.bitbucket.org/2.0/repositories?role=contributor&pagelen=100",
```
Comment (zharinov) [inline lib/platform/bitbucket/__snapshots__/index.spec.ts.snap:1054]: Shouldn't be breaking, test run performed well for me

Comment (viceice) [reply lib/platform/bitbucket/__snapshots__/index.spec.ts.snap:1054]: change should be fine. i think previous url was wrong anyways

File: lib/util/http/github.ts [hunk 5]
```diff
@@ -229,17 +222,15 @@ export class GithubHttp extends Http<GithubHttpOptions, GithubHttpOptions> {
     const path = 'graphql';
```
Comment (viceice) [inline lib/util/http/github.ts:222]: inline this?

**Review by rarkins — CHANGES_REQUESTED** [review]
I just discovered that all `url.*` functions are considered deprecated now: https://nodejs.org/docs/latest-v12.x/api/url.html in favor of the `URL()` constructor which we do use in a few places already. Quite a lot to migrate though.

Can this `resolveBaseUrl()` command use the non-deprecated approach?

File: lib/util/url.ts [hunk 6]
```diff
@@ -1,3 +1,16 @@
...
 export function ensureTrailingSlash(url: string): string {
   return url.replace(/\/?$/, '/');
 }
+
+export function resolveBaseUrl(
+  baseUrl: string,
+  pathOrUrl: string | URL
+): string {
+  const pathOrUrlString = pathOrUrl.toString();
+  const { host, path } = parse(pathOrUrlString);
```
Comment (viceice) [inline lib/util/url.ts:13]:
```suggestion
  const { host, path } = new URL(pathOrUrlString);
```

`URL` is on global object.

**Review by viceice — CHANGES_REQUESTED** [review]
We should check all url imports

**Comment by rarkins** [comment]
Ref: https://github.com/nodejs/node/discussions/23694

**Comment by zharinov** [comment]
- [ ] Make `resolveBaseUrl()` to use non-deprecated URL
- [ ] Search `URL.resolve` calls
- [ ] Group them to separate issues/PRs

**Comment by zharinov** [comment]
Superseded by #7918

**Pull request closed without merging**